marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/builtins/steps/mapping/entity-services/main.sjs

```diff
@@ -0,0 +1,187 @@
+  let id = content.uri;
+  //let's set our output format, so we know what we're exporting
+  let inputFormat = options.inputFormat ? options.inputFormat.toLowerCase() : datahub.flow.consts.DEFAULT_FORMAT;
+  let outputFormat = options.outputFormat ? options.outputFormat.toLowerCase() : datahub.flow.consts.DEFAULT_FORMAT;
+  if (outputFormat !== datahub.flow.consts.JSON && outputFormat !== datahub.flow.consts.XML) {
+    datahub.debug.log({
+      message: 'The output format of type ' + outputFormat + ' is invalid. Valid options are ' + datahub.flow.consts.XML + ' or ' + datahub.flow.consts.JSON + '.',
+      type: 'error'
+    });
+    throw Error('The output format of type ' + outputFormat + ' is invalid. Valid options are ' + datahub.flow.consts.XML + ' or ' + datahub.flow.consts.JSON + '.');
+  }
+
+  let doc = content.value;
+
+  //then we grab our mapping
+  let mappingKey = options.mapping ? `${options.mapping.name}:${options.mapping.version}` : null;
+  let mapping = mappings[mappingKey];
+  if (!mapping && options.mapping && options.mapping.name && options.mapping.version) {
+    let version = parseInt(options.mapping.version);
+    if(isNaN(version)){
+      datahub.debug.log({message: 'Mapping version ('+options.mapping.version+') is invalid.', type: 'error'});
+      throw Error('Mapping version ('+options.mapping.version+') is invalid.');
+    }
+    mapping = defaultLib.getMappingWithVersion(options.mapping.name, version);
+    mappings[mappingKey] = mapping;
+  } else if (!mapping && options.mapping && options.mapping.name) {
+    mapping = defaultLib.getMapping(options.mapping.name);
+    mappings[mappingKey] = mapping;
+  } else if (!mapping) {
+    datahub.debug.log({message: 'You must specify a mapping name.', type: 'error'});
+    throw Error('You must specify a mapping name.');
+  }
+
+  if(fn.empty(mapping)) {
+    let mapError = 'Could not find mapping: ' + options.mapping.name;
+    if (options.mapping.version) {
+      mapError += ' with version #' + options.mapping.version;
+    }
+    datahub.debug.log({message: mapError, type: 'error'});
+    throw Error(mapError);
+  }
+  let mappingURIforXML = fn.replace(xdmp.nodeUri(mapping), 'json$','xml');
+  let targetEntityType = mapping.root.targetEntityType;
+  let entityName = lib.getEntityName(targetEntityType);
+  let entity = lib.getTargetEntity(targetEntityType);
+  let instance;
+  let provenance = {};
+  //Then we obtain the instance and process it from the source context
+  try {
+    instance = es.mapToCanonical(doc, mappingURIforXML, {'entity':entityName,'format':outputFormat});;
+  } catch (e) {
+    datahub.debug.log({message: e, type: 'error'});
+    throw Error(e);
+  }
+
+  content.value = buildEnvelope(entity.info, doc, instance, outputFormat, options);
+  content.provenance = { [content.uri]: provenance };
+  return content;
+}
+
+// Extracted for unit testing purposes
+function buildEnvelope(entityInfo, doc, instance, outputFormat, options) {
```

Comment:
This all looks good, but is it not yet possible to write a unit test for it? Or to setup an integration-level test that sets up an entity and a flow with an ES mapping step and runs it?